<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>SequenceSet - Genomics Documentation</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "SequenceSet";
    var mkdocs_page_input_path = "docs/SequenceSet.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> Genomics Documentation</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">annotations</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../Annotation/">Annotation</a>
                </li>
                <li class="">
                    
    <a class="" href="../AnnotationTable/">AnnotationTable</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">gene</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../Gene/">Gene</a>
                </li>
                <li class="">
                    
    <a class="" href="../GeneTable/">GeneTable</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">region</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../Region/">Region</a>
                </li>
                <li class="">
                    
    <a class="" href="../DisjointRegion/">DisjointRegion</a>
                </li>
                <li class="">
                    
    <a class="" href="../RegionTable/">RegionTable</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">sequence</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../Sequence/">Sequence</a>
                </li>
                <li class="">
                    
    <a class="" href="../DNA/">DNA</a>
                </li>
                <li class="">
                    
    <a class="" href="../RNA/">RNA</a>
                </li>
                <li class="">
                    
    <a class="" href="../AminoAcids/">AminoAcids</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">SequenceSet</a>
    <ul class="subnav">
            
    </ul>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">variant</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../Variant/">Variant</a>
                </li>
                <li class="">
                    
    <a class="" href="../VariantTable/">VariantTable</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">Genomics Documentation</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
        
          <li>sequence &raquo;</li>
        
      
    
    <li>SequenceSet</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="genomics.SequenceSet">SequenceSet</h1>

<pre><code class="python">SequenceSet(self, sequences, start_index=None)
</code></pre>

<p>Abstract class for a collection of sequences.</p>
<p><strong>Properties</strong></p>
<p>id (str): Identifier. If none is provided, uses a uuid.
id_type (str): Source of identifier.
sequences (list of sequences): Genome contents.
start_index (int): TODO
annotations (AnnotationTable): table containing all annotations, created by
    concatenating the individual AnnotationTable for every Sequence in the
    SequenceSet</p>
<h2 id="genomics.SequenceSet.from_file">from_file</h2>

<pre><code class="python">SequenceSet.from_file(filetype=None)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.reindex">reindex</h2>

<pre><code class="python">SequenceSet.reindex(self)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.get_subset">get_subset</h2>

<pre><code class="python">SequenceSet.get_subset(self, slice)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.find">find</h2>

<pre><code class="python">SequenceSet.find(self, query, limit=1)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.count">count</h2>

<pre><code class="python">SequenceSet.count(self, query)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.find_pattern">find_pattern</h2>

<pre><code class="python">SequenceSet.find_pattern(self, pattern, limit=1, overlapping=True)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.count_pattern">count_pattern</h2>

<pre><code class="python">SequenceSet.count_pattern(self, pattern)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.compute_distance">compute_distance</h2>

<pre><code class="python">SequenceSet.compute_distance(self, regionA, regionB, fn='levhenstein')
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.apply_variant">apply_variant</h2>

<pre><code class="python">SequenceSet.apply_variant(self, genomic_variant)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.save_as">save_as</h2>

<pre><code class="python">SequenceSet.save_as(self, path, filetype=None, overwrite_existing=False)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.save_sequences">save_sequences</h2>

<pre><code class="python">SequenceSet.save_sequences(self, path, filetype=None, overwrite_existing=False)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.SequenceSet.save_annotations">save_annotations</h2>

<pre><code class="python">SequenceSet.save_annotations(self, path, filetype=None, overwrite_existing=False)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h1 id="genomics.DNASet">DNASet</h1>

<pre><code class="python">DNASet(self, sequences, start_index=None)
</code></pre>

<p>SequenceSet with DNA-specific methods for fetching and storing</p>
<h2 id="genomics.DNASet.calculate_gc">calculate_gc</h2>

<pre><code class="python">DNASet.calculate_gc(self)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.DNASet.from_db">from_db</h2>

<pre><code class="python">DNASet.from_db(db='ncbi', key_type='accession', with_annotations=True, start_index=1, allowed_chars=None, start_col='start', end_col=('end', 'stop'), key_col=('key', 'accession'), keys_from_file=False, tabular_file=False)
</code></pre>

<p>Fetches multiple sequences or sequence subsets from external db</p>
<p>Compatible with multiple ways of specifying data.
When <code>keys_from_file=False</code>:
- If data_specification is iterable of strings, each string should be
an key for a distinct sequence in the db.
- If data_specification is iterable of tuples, each tuple should be in
the format (key, start, stop) where:
    - key is the db key for the sequence
    - start is the index of the first element to be included
    - end is the index of the last element to be included.
If stop is omitted, uses index of last
element in db sequence. If both start and stop are omitted, entire
sequence will be used.</p>
<p>When <code>keys_from_file=True</code>:
- if data_specification is string, should be a path or url to either:
    - if <code>from_tabular_file=False</code>: a file containing a list of db keys
    - if <code>from_tabular_file=True</code>: a tabular file with a column
    containing db keys, an (optional) column containing start indexes,
    and an (optional) column containing stop indexes
- if data_specification is pandas DataFrame, if should contain a column
    of data keys, (optional) column of start indexes and (optional)
    column of stop indexes</p>
<p><strong>Arguments</strong></p>
<ul>
<li><strong>data_specification (iterable, str, or pd.DataFrame)</strong>: sequences specified
    in the above format</li>
</ul>
<p><strong>Returns</strong></p>
<p><strong>Raises</strong></p>
<ul>
<li><code>ValueError</code>:</li>
</ul>
<h1 id="genomics.RNASet">RNASet</h1>

<pre><code class="python">RNASet(self, sequences, start_index=None)
</code></pre>

<p>SequenceSet with RNA-specific methods for fetching and storing</p>
<h2 id="genomics.RNASet.calculate_gc">calculate_gc</h2>

<pre><code class="python">RNASet.calculate_gc(self)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h2 id="genomics.RNASet.from_db">from_db</h2>

<pre><code class="python">RNASet.from_db(self)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
<h1 id="genomics.AminoAcidsSet">AminoAcidsSet</h1>

<pre><code class="python">AminoAcidsSet(self, sequences, start_index=None)
</code></pre>

<p>SequenceSet with AminoAcids-specific methods for fetching and storing</p>
<h2 id="genomics.AminoAcidsSet.from_db">from_db</h2>

<pre><code class="python">AminoAcidsSet.from_db(self)
</code></pre>

<p><strong>Arguments</strong></p>
<p><strong>Returns</strong></p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../Variant/" class="btn btn-neutral float-right" title="Variant">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../AminoAcids/" class="btn btn-neutral" title="AminoAcids"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../AminoAcids/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../Variant/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>

</body>
</html>
